src/iotjs_string.c

```diff
@@ -98,57 +84,41 @@ bool iotjs_string_is_empty(const iotjs_string_t* str) {
 void iotjs_string_make_empty(iotjs_string_t* str) {
   IOTJS_VALIDATED_STRUCT_METHOD(iotjs_string_t, str);
 
-  if (_this->data != iotjs_empty_string) {
+  if (_this->data != NULL) {
     iotjs_buffer_release(_this->data);
-    _this->cap = _this->size = 0;
-    _this->data = iotjs_empty_string;
+    _this->size = 0;
+    _this->data = NULL;
   }
 }
 
 
-void iotjs_string_append(iotjs_string_t* str, const char* data, int size) {
+void iotjs_string_append(iotjs_string_t* str, const char* data, uint32_t size) {
   IOTJS_VALIDATED_STRUCT_METHOD(iotjs_string_t, str);
 
-  const int kCapacityIncreaseFactor = 2;
-
-  IOTJS_ASSERT(_this->data != NULL);
-  IOTJS_ASSERT(_this->cap >= _this->size);
+  IOTJS_ASSERT(data != NULL);
 
-  if (size < 0) {
-    size = strlen(data);
-  }
   if (size == 0) {
     return;
   }
 
-  if (_this->cap == 0) {
-    // No buffer was allocated.
-    IOTJS_ASSERT(_this->size == 0);
-    IOTJS_ASSERT(_this->data == iotjs_empty_string);
-    _this->cap = _this->size = size;
-    _this->data = iotjs_buffer_allocate(_this->cap + 1);
-    strncpy(_this->data, data, _this->size);
-  } else if (_this->cap >= _this->size + size) {
-    // Have enough capacity to append data.
-    strncpy(_this->data + _this->size, data, size);
-    _this->size += size;
+  if (_this->data != NULL) {
+    _this->data = iotjs_buffer_reallocate(_this->data, _this->size + size);
   } else {
-    // Lack of capacity, calculate next capacity.
-    while (_this->cap < _this->size + size) {
-      _this->cap *= kCapacityIncreaseFactor;
-    }
-    // Reallocate buffer and copy data.
-    _this->data = iotjs_buffer_reallocate(_this->data, _this->cap + 1);
-    strncpy(_this->data + _this->size, data, size);
-    _this->size += size;
+    IOTJS_ASSERT(_this->size == 0);
+    _this->data = iotjs_buffer_allocate(size);
   }
-  _this->data[_this->size] = 0;
+
+  strncpy(_this->data + _this->size, data, size);
```

Comment:
The old code used `strncpy` too. I think it is safer to use `strncpy`, because it adds zero padding if the size is bigger.